esphome/components/tcs34725/tcs34725.cpp

```diff
@@ -226,14 +229,84 @@ void TCS34725Component::update() {
+  
+  if (integration_time_auto_){
+    //change integration time an gain to achieve maximum resolution an dynamic range
+    //calculate optimal integration time to achieve 70% satuaration
+    float integration_time_ideal;
+    integration_time_ideal = 60 / ((float) raw_c /655.35) * this->integration_time_;
+
+    uint8_t gainRegVal = this->gain_reg_; 
+    // increase gain if less than 20% of white channel used and high integration time
+    // increase only if not already maximum
+    // do not use max gain, as ist will not get better
+    if (this->gain_reg_ < 2){
+      if ( ((float) raw_c /655.35 < 20.f) && (this->integration_time_ > 600.f)){
+
+        gainRegVal= this->gain_reg_ +1;
+        // update integration time to new situation
+        integration_time_ideal = integration_time_ideal /4;
+        //ESP_LOGD(TAG, "TCS34725I Gain increase to %d", 2^gainRegVal);
+      }
+    }
+
+    // decrease gain, if very high clear values and integration times alreadey low
+    if (this->gain_reg_ > 0){
+      if ( 70 < ((float) raw_c /655.35) && (this->integration_time_ < 200)){
 
+        gainRegVal = this->gain_reg_ - 1;
+        // update integration time to new situation
+        integration_time_ideal = integration_time_ideal *4;
+        //ESP_LOGD(TAG, "TCS34725I Gain adjust to %d", 2^gainRegVal);
+      }
+    }    
+
+    // saturate integration times
+    float integration_time_next = integration_time_ideal;
+    if (integration_time_ideal> 2.4f*256){
+      integration_time_next = 2.4f*256;
+    } 
+    if (integration_time_ideal < 154){
+      integration_time_next = 154;
+    } 
+
+    // calculate register value from timing 
+    uint8_t regvalATIME = (uint8_t) (256.f - integration_time_next / 2.4f);
+    uint8_t actual_gain = pow(4, this->gain_reg_);
+    ESP_LOGI(TAG, "Integration time: %.1fms, ideal: %.1fms regval_new %d Gain_reg: %d",this->integration_time_ , integration_time_next, regvalATIME, actual_gain);
+  
+    if (this->integration_reg_ != regvalATIME  || gainRegVal != this->gain_reg_){
+      this->integration_reg_ = regvalATIME;
+      this->gain_reg_ = gainRegVal;
+      if (this->write_config_register_(TCS34725_REGISTER_ATIME, this->integration_reg_) != i2c::ERROR_OK ||
+          this->write_config_register_(TCS34725_REGISTER_CONTROL, this->gain_reg_) != i2c::ERROR_OK) {
+        this->mark_failed();
+        ESP_LOGW(TAG, "TCS34725I update timing failed!");
+      } else {
+        this->integration_time_ = integration_time_next;
+      }
+    }
+}
   this->status_clear_warning();
 }
 void TCS34725Component::set_integration_time(TCS34725IntegrationTime integration_time) {
-  this->integration_reg_ = integration_time;
-  this->integration_time_ = (256.f - integration_time) * 2.4f;
+  // if an integration time is 0x100, this is auto start with 154ms as this gives best starting point
+  TCS34725IntegrationTime my_integration_time_regval;
+
+  if (integration_time == 0x100){
+    integration_time_auto_ = true;
```

Comment:
```suggestion
    this->integration_time_auto_ = true;
```
```suggestion
    this->integration_time_auto_ = false;
```

See https://esphome.io/guides/contributing.html#codebase-standards